src/parse.c

```diff
@@ -97,6 +97,20 @@ static bool           translate_on = true;
 static nametab_t     *nametab = NULL;
 static bool           bootstrapping = false;
 
+typedef struct _label_cnts label_cnts_t;
+
+struct _label_cnts {
+   int proc;
+   int loop;
+   int stmt;
+};
```

Comment:
The counters should go in the `scope_t` struct in `names.c` rather than here as that already tracks information for each scope.

Reply:
Done, I moved all counters to be per-scope in `scope_t`. It simplifies the parsing as it avoids state dependency
within the parser itself. OTOH, now each scope has such counters, which leads to more memory usage. Further,
some of the counters are unused, leading to waste of memory (e.g. each loop scope has loop counter, but it
is unused by numbering since loops are using counter from surrounding process or subprogram). 

It might be improved by keeping only two counters since per scope:
1. Entity, architecture scope contains only statements to be converted to processes, so counter for _P is enough.
2. Scopes within subprogram or process do not require processes to be counted, so they only need two distinct counters for _L and _S.

Note that the modified code now does not linearly label statements within loop scopes of processes, it restarts
the counter again.

Reply:
Btw. how do I run only particular test?

Reply:
> OTOH, now each scope has such counters, which leads to more memory usage. Further, some of the counters are unused, leading to waste of memory (e.g. each loop scope has loop counter, but it is unused by numbering since loops are using counter from surrounding process or subprogram).

Is it significant? Those counters are 12 bytes per scope. The scopes are freed when the parser exits them. What's a reasonable worst-case nesting depth - 10, 20? 240 bytes is nothing.

> 
> Note that the modified code now does not linearly label statements within loop scopes of processes, it restarts the counter again.

I don't think it matters: the LRM doesn't specify these labels (?) because you can't use them in an selected name.

> Btw. how do I run only particular test?

Easiest way is

```
./bin/run_regr <test name or sub-string>
```

Reply:
Ok, that makes sense, thanks, I did not realize scopes are freed when parser exits them.
As for the statements, it does not matter at all as LRM says nothing there.

Reply:
The scope structure is just used to hold some temporary data during symbol lookup and overload resolution. After parsing all the relevant data is stored in the AST (e.g. `tree_ref`)